packages/core/src/indexing/sqlite/chronological-order.ts

```diff
@@ -0,0 +1,184 @@
+  { column: 'last_anchored_at', order: 'DESC', nulls: 'last' }, // (`last_anchored_at` is null) DESC
+  { column: 'last_anchored_at', order: 'DESC' },
+  { column: 'created_at', order: 'DESC' },
+]
+
+/**
+ * Reverse ASC to DESC, and DESC to ASC in an order clause.
+ */
+function reverseOrder<T extends { order: string }>(entries: Array<T>): Array<T> {
+  const reverse = {
+    ASC: 'DESC',
+    DESC: 'ASC',
+  }
+  return entries.map((entry) => ({ ...entry, order: reverse[entry.order] }))
+}
+
+export class ChronologicalOrder {
+  constructor(private readonly dataSource: DataSource, private readonly knexConnection: Knex) {}
+
+  async page(query: BaseQuery & Pagination): Promise<Page<StreamID>> {
+    const pagination = parsePagination(query)
+    const paginationKind = pagination.kind
+    switch (paginationKind) {
+      case PaginationKind.FORWARD: {
+        const limit = pagination.first
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalForwardQuery(query, pagination)
+        )
+        const entries = response.slice(0, limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: response.length > limit,
+            hasPreviousPage: false,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      case PaginationKind.BACKWARD: {
+        const limit = pagination.last
+        const response = await this.query<Array<Selected>>(
+          this.chronologicalBackwardQuery(query, pagination)
+        )
+        const entries = response.slice(-limit)
+        const firstEntry = entries[0]
+        const lastEntry = entries[entries.length - 1]
+        return {
+          entries: entries.map((row) => StreamID.fromString(row.stream_id)),
+          pageInfo: {
+            hasNextPage: false,
+            hasPreviousPage: response.length > limit,
+            endCursor: Cursor.stringify(asChronologicalCursor(lastEntry)),
+            startCursor: Cursor.stringify(asChronologicalCursor(firstEntry)),
+          },
+        }
+      }
+      default:
+        throw new UnsupportedOrderingError(paginationKind)
+    }
+  }
+
+  private chronologicalForwardQuery(
+    query: BaseQuery,
+    pagination: ForwardPaginationQuery
+  ): Knex.QueryBuilder {
+    const tableName = asTableName(query.model)
+    const base = this.knexConnection
+      .from(tableName)
+      .select('stream_id', 'last_anchored_at', 'created_at')
+      .orderBy(CHRONOLOGICAL_ORDER)
+      .limit(pagination.first + 1)
+    if (pagination.after) {
+      const after = Cursor.parse(pagination.after)
+      if (after.last_anchored_at) {
+        return base.where('last_anchored_at', '<', after.last_anchored_at)
```

Comment:
I'm having a hard time making sense of the sorting logic here.  Will revisit after comments are added explaining which of ForwardQuery or BackwardQuery is supposed to get the most recent entries or the least recent entries

Reply:
Okay, reading more I think I mostly get it now.  I think ForwardQuery is supposed to get the most recent entries and return them in decreasing recency order, while BackwardQuery gets the oldest entries but the results get newer as you iterate.  I was a bit thrown by ForwardCursor using `after` for subsequent batches given that it's actually requesting older entries, but I guess that's because the Page API is general and doesn't assume the query is in chronological order, so it's "after" the previous batch, but in this case that corresponds to entries created "before" on the network.

Assuming I have that all right, my only remaining question is why `CHRONOLOGICAL_ORDER` says `last_anchored_at` should sort `nulls: 'last'`.  If you're trying to get the most recent entries, the unanchored entries should come first, so then shouldn't it be `nulls: 'first'` in `CHRONOLOGICAL_ORDER`?

I see your unit test CSV file has the order I expect, so I'm sure it actually works, but I'm just confused as to how it's working.